File: scripts/extract-edge-function-metadata.py
#!/usr/bin/env python3
"""
Edge Function Metadata Extraction Script
Feature: 019-create-comprehensive-edge-functions-api-reference

Analyzes all Edge Functions in supabase/functions/ to extract:
- Function name and path
- HTTP methods supported
- Request/response interfaces
- JSDoc comments and API documentation
- Dependencies on _shared utilities
- Authentication requirements

Usage:
  python3 scripts/extract-edge-function-metadata.py [options]

Options:
  --help              Show this help message
  --output <file>     Write JSON output to file (default: stdout)
  --validate          Validate all functions were successfully parsed
  --verbose           Show detailed parsing information
"""

import os
import sys
import json
import re
import argparse
from pathlib import Path
from datetime import datetime
from typing import List, Dict, Any, Optional, Set


class ParameterInfo:
    def __init__(self, name: str, description: str = "", required: bool = False, param_type: str = ""):
        self.name = name
        self.type = param_type
        self.required = required
        self.description = description

    def to_dict(self):
        return {
            "name": self.name,
            "type": self.type,
            "required": self.required,
            "description": self.description
        }


class PropertyInfo:
    def __init__(self, name: str, prop_type: str, optional: bool = False):
        self.name = name
        self.type = prop_type
        self.optional = optional

    def to_dict(self):
        return {
            "name": self.name,
            "type": self.type,
            "optional": self.optional
        }


class InterfaceInfo:
    def __init__(self, name: str, properties: List[PropertyInfo], line_number: int):
        self.name = name
        self.properties = properties
        self.line_number = line_number

    def to_dict(self):
        return {
            "name": self.name,
            "properties": [prop.to_dict() for prop in self.properties],
            "lineNumber": self.line_number
        }


class FunctionMetadata:
    def __init__(self, name: str, path: str, relative_path: str):
        self.name = name
        self.path = path
        self.relative_path = relative_path
        self.http_methods: List[str] = []
        self.description: Optional[str] = None
        self.endpoint: Optional[str] = None
        self.query_params: List[ParameterInfo] = []
        self.body_params: List[ParameterInfo] = []
        self.response_type: Optional[str] = None
        self.interfaces: List[InterfaceInfo] = []
        self.shared_dependencies: List[str] = []
        self.has_auth = False
        self.has_cors = False
        self.has_rate_limit = False
        self.jsdoc_comments: List[str] = []
        self.feature: Optional[str] = None
        self.errors: List[str] = []

    def to_dict(self):
        return {
            "name": self.name,
            "path": self.path,
            "relativePath": self.relative_path,
            "httpMethods": self.http_methods,
            "description": self.description,
            "endpoint": self.endpoint,
            "queryParams": [p.to_dict() for p in self.query_params],
            "bodyParams": [p.to_dict() for p in self.body_params],
            "responseType": self.response_type,
            "interfaces": [i.to_dict() for i in self.interfaces],
            "sharedDependencies": self.shared_dependencies,
            "hasAuth": self.has_auth,
            "hasCors": self.has_cors,
            "hasRateLimit": self.has_rate_limit,
            "jsdocComments": self.jsdoc_comments,
            "feature": self.feature,
            "errors": self.errors
        }


def extract_jsdoc_comments(content: str) -> List[str]:
    """Extract JSDoc comments from file content"""
    jsdoc_regex = re.compile(r'/\*\*([\s\S]*?)\*/', re.MULTILINE)
    matches = jsdoc_regex.findall(content)
    return [match.strip() for match in matches]


def extract_endpoint(jsdoc_comments: List[str]) -> Optional[str]:
    """Extract endpoint information from JSDoc comments"""
    for comment in jsdoc_comments:
        match = re.search(r'@endpoint\s+(GET|POST|PUT|PATCH|DELETE)\s+(.+)', comment)
        if match:
            return f"{match.group(1)} {match.group(2).strip()}"
    return None


def extract_description(jsdoc_comments: List[str]) -> Optional[str]:
    """Extract description from JSDoc comments"""
    if not jsdoc_comments:
        return None

    # First comment block, before any @tags
    first_comment = jsdoc_comments[0]
    lines = first_comment.split('\n')
    description_lines = []

    for line in lines:
        trimmed = re.sub(r'^\s*\*\s?', '', line).strip()
        if trimmed.startswith('@'):
            break
        if trimmed:
            description_lines.append(trimmed)

    return ' '.join(description_lines) if description_lines else None


def extract_feature(jsdoc_comments: List[str]) -> Optional[str]:
    """Extract feature tag from JSDoc comments"""
    for comment in jsdoc_comments:
        match = re.search(r'Feature:\s*([^\n]+)', comment, re.IGNORECASE)
        if match:
            return match.group(1).strip()
    return None


def extract_query_params(jsdoc_comments: List[str]) -> List[ParameterInfo]:
    """Extract query parameters from JSDoc comments"""
    params = []

    for comment in jsdoc_comments:
        lines = comment.split('\n')
        for line in lines:
            match = re.search(r'@query\s+(\w+)\s*-\s*(.+?)(?:\s*\((required|optional)\))?$', line, re.IGNORECASE)
            if match:
                name = match.group(1)
                description = match.group(2).strip()
                requirement = match.group(3)
                required = (requirement and requirement.lower() == 'required') or '(required)' in description
                params.append(ParameterInfo(name, description, required))

    return params


def extract_http_methods(content: str) -> List[str]:
    """Extract HTTP methods from code"""
    methods: Set[str] = set()

    # Check for explicit method checks
    method_checks = re.finditer(r"req\.method\s*===\s*['\"](\w+)['\"]", content)
    for match in method_checks:
        methods.add(match.group(1))

    # Check for method validation
    allowed_methods = re.finditer(r"allowed.*methods?.*['\"](\w+)['\"]", content, re.IGNORECASE)
    for match in allowed_methods:
        methods.add(match.group(1).upper())

    # If no methods found but serve() exists, check JSDoc for method
    if not methods and 'serve(' in content:
        endpoint_match = re.search(r'@endpoint\s+(GET|POST|PUT|PATCH|DELETE)', content)
        if endpoint_match:
            methods.add(endpoint_match.group(1))

    # Remove OPTIONS as it's just for CORS
    methods.discard('OPTIONS')

    return sorted(list(methods))


def extract_interfaces(content: str) -> List[InterfaceInfo]:
    """Extract TypeScript interfaces from code"""
    interfaces = []
    interface_regex = re.compile(r'interface\s+(\w+)\s*\{([^}]+)\}', re.MULTILINE)

    for match in interface_regex.finditer(content):
        name = match.group(1)
        body = match.group(2)
        properties = []

        # Parse interface properties
        property_lines = body.split('\n')
        for line in property_lines:
            prop_match = re.match(r'^\s*(\w+)(\?)?:\s*(.+?);?\s*$', line)
            if prop_match:
                prop_name = prop_match.group(1)
                optional = prop_match.group(2) == '?'
                prop_type = prop_match.group(3).strip()
                properties.append(PropertyInfo(prop_name, prop_type, optional))

        # Find line number
        line_number = content[:match.start()].count('\n') + 1

        interfaces.append(InterfaceInfo(name, properties, line_number))

    return interfaces


def extract_shared_dependencies(content: str) -> List[str]:
    """Extract shared dependencies"""
    dependencies: Set[str] = set()
    import_regex = re.compile(r"from\s+['\"]\.\.?\/_shared\/([^'\"]+)['\"]")

    for match in import_regex.finditer(content):
        dep = match.group(1).replace('.ts', '')
        dependencies.add(dep)

    return sorted(list(dependencies))


def has_authentication(content: str) -> bool:
    """Check for authentication requirements"""
    return any([
        'auth.getUser()' in content,
        'validateJWT' in content,
        'createUserClient' in content,
        'Authorization' in content
    ])


def has_cors(content: str) -> bool:
    """Check for CORS headers"""
    return 'corsHeaders' in content or 'CORS' in content


def has_rate_limiting(content: str) -> bool:
    """Check for rate limiting"""
    return any([
        'rate-limit' in content,
        'rateLimit' in content,
        'RateLimiter' in content
    ])


def extract_function_metadata(function_path: Path, function_name: str, base_path: Path) -> FunctionMetadata:
    """Extract metadata from a single Edge Function"""
    relative_path = str(function_path.relative_to(base_path))
    metadata = FunctionMetadata(function_name, str(function_path), relative_path)

    try:
        # Read index.ts file
        index_path = function_path / 'index.ts'
        content = index_path.read_text(encoding='utf-8')

        # Extract JSDoc comments
        metadata.jsdoc_comments = extract_jsdoc_comments(content)

        # Extract endpoint and description
        metadata.endpoint = extract_endpoint(metadata.jsdoc_comments)
        metadata.description = extract_description(metadata.jsdoc_comments)
        metadata.feature = extract_feature(metadata.jsdoc_comments)

        # Extract parameters
        metadata.query_params = extract_query_params(metadata.jsdoc_comments)

        # Extract HTTP methods
        metadata.http_methods = extract_http_methods(content)

        # Extract interfaces
        metadata.interfaces = extract_interfaces(content)

        # Extract dependencies
        metadata.shared_dependencies = extract_shared_dependencies(content)

        # Check for features
        metadata.has_auth = has_authentication(content)
        metadata.has_cors = has_cors(content)
        metadata.has_rate_limit = has_rate_limiting(content)

    except Exception as e:
        metadata.errors.append(f"Failed to parse: {str(e)}")

    return metadata


def extract_all_metadata(functions_dir: Path, verbose: bool = False) -> Dict[str, Any]:
    """Extract metadata from all Edge Functions"""
    functions = []
    successfully_parsed = 0
    failed = 0

    # Walk through functions directory
    for entry in sorted(functions_dir.iterdir()):
        if not entry.is_dir() or entry.name.startswith('_'):
            continue

        if verbose:
            print(f"Processing: {entry.name}", file=sys.stderr)

        metadata = extract_function_metadata(entry, entry.name, functions_dir)
        functions.append(metadata)

        if not metadata.errors:
            successfully_parsed += 1
        else:
            failed += 1
            if verbose:
                print(f"  Errors: {', '.join(metadata.errors)}", file=sys.stderr)

    # Generate summary statistics
    by_method: Dict[str, int] = {}
    by_feature: Dict[str, int] = {}
    with_auth = 0
    with_rate_limit = 0
    total_interfaces = 0

    for func in functions:
        # Count by method
        for method in func.http_methods:
            by_method[method] = by_method.get(method, 0) + 1

        # Count by feature
        if func.feature:
            by_feature[func.feature] = by_feature.get(func.feature, 0) + 1

        # Count auth and rate limiting
        if func.has_auth:
            with_auth += 1
        if func.has_rate_limit:
            with_rate_limit += 1

        # Count interfaces
        total_interfaces += len(func.interfaces)

    return {
        "totalFunctions": len(functions),
        "successfullyParsed": successfully_parsed,
        "failed": failed,
        "functions": [f.to_dict() for f in sorted(functions, key=lambda x: x.name)],
        "extractedAt": datetime.utcnow().isoformat() + 'Z',
        "summary": {
            "byMethod": by_method,
            "byFeature": by_feature,
            "withAuth": with_auth,
            "withRateLimit": with_rate_limit,
            "totalInterfaces": total_interfaces
        }
    }


def main():
    parser = argparse.ArgumentParser(
        description='Extract metadata from Supabase Edge Functions for API documentation.'
    )
    parser.add_argument('--output', '-o', type=str, help='Write JSON output to file (default: stdout)')
    parser.add_argument('--validate', action='store_true', help='Validate all functions were successfully parsed')
    parser.add_argument('--verbose', '-v', action='store_true', help='Show detailed parsing information')

    args = parser.parse_args()

    functions_dir = Path('./supabase/functions')

    # Check if directory exists
    if not functions_dir.exists():
        print(f"Error: Directory not found: {functions_dir}", file=sys.stderr)
        print("Make sure you are running this script from the project root.", file=sys.stderr)
        sys.exit(1)

    if args.verbose:
        print('Extracting metadata from Edge Functions...\n', file=sys.stderr)

    # Extract metadata
    result = extract_all_metadata(functions_dir, args.verbose)

    # Validate if requested
    if args.validate:
        print('\n=== Validation Results ===', file=sys.stderr)
        print(f"Total functions: {result['totalFunctions']}", file=sys.stderr)
        print(f"Successfully parsed: {result['successfullyParsed']}", file=sys.stderr)
        print(f"Failed: {result['failed']}", file=sys.stderr)

        if result['failed'] > 0:
            print('\nFunctions with errors:', file=sys.stderr)
            for func in result['functions']:
                if func['errors']:
                    print(f"  - {func['name']}: {', '.join(func['errors'])}", file=sys.stderr)
            sys.exit(1)
        else:
            print('\n✓ All functions successfully parsed', file=sys.stderr)
            sys.exit(0)

    # Output JSON
    json_output = json.dumps(result, indent=2)

    if args.output:
        output_path = Path(args.output)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(json_output, encoding='utf-8')
        if args.verbose:
            print(f"\n✓ Metadata written to: {args.output}", file=sys.stderr)
    else:
        print(json_output)

    # Summary in verbose mode
    if args.verbose:
        print('\n=== Summary ===', file=sys.stderr)
        print(f"Total functions: {result['totalFunctions']}", file=sys.stderr)
        print(f"Successfully parsed: {result['successfullyParsed']}", file=sys.stderr)
        print(f"HTTP Methods: {json.dumps(result['summary']['byMethod'], indent=2)}", file=sys.stderr)
        print(f"With Auth: {result['summary']['withAuth']}", file=sys.stderr)
        print(f"With Rate Limiting: {result['summary']['withRateLimit']}", file=sys.stderr)
        print(f"Total Interfaces: {result['summary']['totalInterfaces']}", file=sys.stderr)


if __name__ == '__main__':
    main()

File: scripts/extract-edge-function-metadata.ts
#!/usr/bin/env -S deno run --allow-read
/**
 * Edge Function Metadata Extraction Script
 * Feature: 019-create-comprehensive-edge-functions-api-reference
 *
 * Analyzes all Edge Functions in supabase/functions/ to extract:
 * - Function name and path
 * - HTTP methods supported
 * - Request/response interfaces
 * - JSDoc comments and API documentation
 * - Dependencies on _shared utilities
 * - Authentication requirements
 *
 * Usage:
 *   deno run --allow-read scripts/extract-edge-function-metadata.ts [options]
 *
 * Options:
 *   --help              Show this help message
 *   --output <file>     Write JSON output to file (default: stdout)
 *   --validate          Validate all functions were successfully parsed
 *   --verbose           Show detailed parsing information
 */

import { parse } from 'https://deno.land/std@0.208.0/flags/mod.ts';
import { walk } from 'https://deno.land/std@0.208.0/fs/walk.ts';
import { join, relative } from 'https://deno.land/std@0.208.0/path/mod.ts';

// ============================================================================
// Types
// ============================================================================

interface FunctionMetadata {
  name: string;
  path: string;
  relativePath: string;
  httpMethods: string[];
  description?: string;
  endpoint?: string;
  queryParams: ParameterInfo[];
  bodyParams: ParameterInfo[];
  responseType?: string;
  interfaces: InterfaceInfo[];
  sharedDependencies: string[];
  hasAuth: boolean;
  hasCors: boolean;
  hasRateLimit: boolean;
  jsdocComments: string[];
  feature?: string;
  errors: string[];
}

interface ParameterInfo {
  name: string;
  type?: string;
  required: boolean;
  description?: string;
}

interface InterfaceInfo {
  name: string;
  properties: PropertyInfo[];
  lineNumber: number;
}

interface PropertyInfo {
  name: string;
  type: string;
  optional: boolean;
}

interface ExtractionResult {
  totalFunctions: number;
  successfullyParsed: number;
  failed: number;
  functions: FunctionMetadata[];
  extractedAt: string;
  summary: {
    byMethod: Record<string, number>;
    byFeature: Record<string, number>;
    withAuth: number;
    withRateLimit: number;
    totalInterfaces: number;
  };
}

// ============================================================================
// CLI Argument Parsing
// ============================================================================

const args = parse(Deno.args, {
  boolean: ['help', 'validate', 'verbose'],
  string: ['output'],
  alias: {
    h: 'help',
    o: 'output',
    v: 'verbose',
  },
});

if (args.help) {
  console.log(`
Usage: extract-edge-function-metadata [options]

Extract metadata from Supabase Edge Functions for API documentation.

Options:
  -h, --help              Show this help message
  -o, --output <file>     Write JSON output to file (default: stdout)
      --validate          Validate all functions were successfully parsed
  -v, --verbose           Show detailed parsing information

Examples:
  # Display metadata as JSON
  deno run --allow-read scripts/extract-edge-function-metadata.ts

  # Save to file
  deno run --allow-read scripts/extract-edge-function-metadata.ts --output docs/api/.metadata/functions-inventory.json

  # Validate extraction
  deno run --allow-read scripts/extract-edge-function-metadata.ts --validate
  `);
  Deno.exit(0);
}

// ============================================================================
// Metadata Extraction Functions
// ============================================================================

/**
 * Extract JSDoc comments from file content
 */
function extractJSDocComments(content: string): string[] {
  const jsdocRegex = /\/\*\*([\s\S]*?)\*\//g;
  const comments: string[] = [];
  let match;

  while ((match = jsdocRegex.exec(content)) !== null) {
    comments.push(match[1].trim());
  }

  return comments;
}

/**
 * Extract endpoint information from JSDoc comments
 */
function extractEndpoint(jsdocComments: string[]): string | undefined {
  for (const comment of jsdocComments) {
    const endpointMatch = comment.match(/@endpoint\s+(GET|POST|PUT|PATCH|DELETE)\s+(.+)/);
    if (endpointMatch) {
      return `${endpointMatch[1]} ${endpointMatch[2].trim()}`;
    }
  }
  return undefined;
}

/**
 * Extract description from JSDoc comments
 */
function extractDescription(jsdocComments: string[]): string | undefined {
  if (jsdocComments.length === 0) return undefined;

  // First comment block, before any @tags
  const firstComment = jsdocComments[0];
  const lines = firstComment.split('\n');
  const descriptionLines: string[] = [];

  for (const line of lines) {
    const trimmed = line.replace(/^\s*\*\s?/, '').trim();
    if (trimmed.startsWith('@')) break;
    if (trimmed) descriptionLines.push(trimmed);
  }

  return descriptionLines.length > 0 ? descriptionLines.join(' ') : undefined;
}

/**
 * Extract feature tag from JSDoc comments
 */
function extractFeature(jsdocComments: string[]): string | undefined {
  for (const comment of jsdocComments) {
    const featureMatch = comment.match(/Feature:\s*([^\n]+)/i);
    if (featureMatch) {
      return featureMatch[1].trim();
    }
  }
  return undefined;
}

/**
 * Extract query parameters from JSDoc comments
 */
function extractQueryParams(jsdocComments: string[]): ParameterInfo[] {
  const params: ParameterInfo[] = [];

  for (const comment of jsdocComments) {
    const lines = comment.split('\n');
    for (const line of lines) {
      const queryMatch = line.match(/@query\s+(\w+)\s*-\s*(.+?)(?:\s*\((required|optional)\))?$/i);
      if (queryMatch) {
        const [, name, description, requirement] = queryMatch;
        params.push({
          name,
          description: description.trim(),
          required: requirement?.toLowerCase() === 'required' || description.includes('(required)'),
        });
      }
    }
  }

  return params;
}

/**
 * Extract HTTP methods from code
 */
function extractHTTPMethods(content: string): string[] {
  const methods = new Set<string>();

  // Check for explicit method checks
  const methodChecks = content.matchAll(/req\.method\s*===\s*['"](\w+)['"]/g);
  for (const match of methodChecks) {
    methods.add(match[1]);
  }

  // Check for method validation
  const allowedMethods = content.matchAll(/allowed.*methods?.*['"](\w+)['"]/gi);
  for (const match of allowedMethods) {
    methods.add(match[1].toUpperCase());
  }

  // If no methods found but serve() exists, assume it handles the method internally
  if (methods.size === 0 && content.includes('serve(')) {
    // Check endpoint JSDoc for method
    const endpointMatch = content.match(/@endpoint\s+(GET|POST|PUT|PATCH|DELETE)/);
    if (endpointMatch) {
      methods.add(endpointMatch[1]);
    }
  }

  // Remove OPTIONS as it's just for CORS
  methods.delete('OPTIONS');

  return Array.from(methods);
}

/**
 * Extract TypeScript interfaces from code
 */
function extractInterfaces(content: string): InterfaceInfo[] {
  const interfaces: InterfaceInfo[] = [];
  const interfaceRegex = /interface\s+(\w+)\s*\{([^}]+)\}/g;
  let match;

  while ((match = interfaceRegex.exec(content)) !== null) {
    const [, name, body] = match;
    const properties: PropertyInfo[] = [];

    // Parse interface properties
    const propertyLines = body.split('\n');
    for (const line of propertyLines) {
      const propMatch = line.match(/^\s*(\w+)(\?)?:\s*(.+?);?\s*$/);
      if (propMatch) {
        const [, propName, optional, type] = propMatch;
        properties.push({
          name: propName,
          type: type.trim(),
          optional: optional === '?',
        });
      }
    }

    // Find line number
    const lineNumber = content.substring(0, match.index).split('\n').length;

    interfaces.push({ name, properties, lineNumber });
  }

  return interfaces;
}

/**
 * Extract shared dependencies
 */
function extractSharedDependencies(content: string): string[] {
  const dependencies = new Set<string>();
  const importRegex = /from\s+['"]\.\.?\/_shared\/([^'"]+)['"]/g;
  let match;

  while ((match = importRegex.exec(content)) !== null) {
    dependencies.add(match[1].replace('.ts', ''));
  }

  return Array.from(dependencies);
}

/**
 * Check for authentication requirements
 */
function hasAuthentication(content: string): boolean {
  return (
    content.includes('auth.getUser()') ||
    content.includes('validateJWT') ||
    content.includes('createUserClient') ||
    content.includes('Authorization')
  );
}

/**
 * Check for CORS headers
 */
function hasCORS(content: string): boolean {
  return content.includes('corsHeaders') || content.includes('CORS');
}

/**
 * Check for rate limiting
 */
function hasRateLimiting(content: string): boolean {
  return (
    content.includes('rate-limit') ||
    content.includes('rateLimit') ||
    content.includes('RateLimiter')
  );
}

/**
 * Extract metadata from a single Edge Function
 */
async function extractFunctionMetadata(
  functionPath: string,
  functionName: string,
  basePath: string
): Promise<FunctionMetadata> {
  const metadata: FunctionMetadata = {
    name: functionName,
    path: functionPath,
    relativePath: relative(basePath, functionPath),
    httpMethods: [],
    queryParams: [],
    bodyParams: [],
    interfaces: [],
    sharedDependencies: [],
    hasAuth: false,
    hasCors: false,
    hasRateLimit: false,
    jsdocComments: [],
    errors: [],
  };

  try {
    // Read index.ts file
    const indexPath = join(functionPath, 'index.ts');
    const content = await Deno.readTextFile(indexPath);

    // Extract JSDoc comments
    metadata.jsdocComments = extractJSDocComments(content);

    // Extract endpoint and description
    metadata.endpoint = extractEndpoint(metadata.jsdocComments);
    metadata.description = extractDescription(metadata.jsdocComments);
    metadata.feature = extractFeature(metadata.jsdocComments);

    // Extract parameters
    metadata.queryParams = extractQueryParams(metadata.jsdocComments);

    // Extract HTTP methods
    metadata.httpMethods = extractHTTPMethods(content);

    // Extract interfaces
    metadata.interfaces = extractInterfaces(content);

    // Extract dependencies
    metadata.sharedDependencies = extractSharedDependencies(content);

    // Check for features
    metadata.hasAuth = hasAuthentication(content);
    metadata.hasCors = hasCORS(content);
    metadata.hasRateLimit = hasRateLimiting(content);
  } catch (error) {
    metadata.errors.push(`Failed to parse: ${error.message}`);
  }

  return metadata;
}

/**
 * Extract metadata from all Edge Functions
 */
async function extractAllMetadata(
  functionsDir: string
): Promise<ExtractionResult> {
  const functions: FunctionMetadata[] = [];
  let successfullyParsed = 0;
  let failed = 0;

  // Walk through functions directory
  for await (const entry of Deno.readDir(functionsDir)) {
    if (!entry.isDirectory || entry.name.startsWith('_')) {
      continue;
    }

    const functionPath = join(functionsDir, entry.name);

    if (args.verbose) {
      console.error(`Processing: ${entry.name}`);
    }

    const metadata = await extractFunctionMetadata(
      functionPath,
      entry.name,
      functionsDir
    );

    functions.push(metadata);

    if (metadata.errors.length === 0) {
      successfullyParsed++;
    } else {
      failed++;
      if (args.verbose) {
        console.error(`  Errors: ${metadata.errors.join(', ')}`);
      }
    }
  }

  // Generate summary statistics
  const byMethod: Record<string, number> = {};
  const byFeature: Record<string, number> = {};
  let withAuth = 0;
  let withRateLimit = 0;
  let totalInterfaces = 0;

  for (const func of functions) {
    // Count by method
    for (const method of func.httpMethods) {
      byMethod[method] = (byMethod[method] || 0) + 1;
    }

    // Count by feature
    if (func.feature) {
      byFeature[func.feature] = (byFeature[func.feature] || 0) + 1;
    }

    // Count auth and rate limiting
    if (func.hasAuth) withAuth++;
    if (func.hasRateLimit) withRateLimit++;

    // Count interfaces
    totalInterfaces += func.interfaces.length;
  }

  return {
    totalFunctions: functions.length,
    successfullyParsed,
    failed,
    functions: functions.sort((a, b) => a.name.localeCompare(b.name)),
    extractedAt: new Date().toISOString(),
    summary: {
      byMethod,
      byFeature,
      withAuth,
      withRateLimit,
      totalInterfaces,
    },
  };
}

// ============================================================================
// Main Execution
// ============================================================================

async function main() {
  const functionsDir = './supabase/functions';

  // Check if directory exists
  try {
    await Deno.stat(functionsDir);
  } catch {
    console.error(`Error: Directory not found: ${functionsDir}`);
    console.error('Make sure you are running this script from the project root.');
    Deno.exit(1);
  }

  if (args.verbose) {
    console.error('Extracting metadata from Edge Functions...\n');
  }

  // Extract metadata
  const result = await extractAllMetadata(functionsDir);

  // Validate if requested
  if (args.validate) {
    console.error('\n=== Validation Results ===');
    console.error(`Total functions: ${result.totalFunctions}`);
    console.error(`Successfully parsed: ${result.successfullyParsed}`);
    console.error(`Failed: ${result.failed}`);

    if (result.failed > 0) {
      console.error('\nFunctions with errors:');
      for (const func of result.functions) {
        if (func.errors.length > 0) {
          console.error(`  - ${func.name}: ${func.errors.join(', ')}`);
        }
      }
      Deno.exit(1);
    } else {
      console.error('\n✓ All functions successfully parsed');
      Deno.exit(0);
    }
  }

  // Output JSON
  const jsonOutput = JSON.stringify(result, null, 2);

  if (args.output) {
    await Deno.writeTextFile(args.output, jsonOutput);
    if (args.verbose) {
      console.error(`\n✓ Metadata written to: ${args.output}`);
    }
  } else {
    console.log(jsonOutput);
  }

  // Summary in verbose mode
  if (args.verbose) {
    console.error('\n=== Summary ===');
    console.error(`Total functions: ${result.totalFunctions}`);
    console.error(`Successfully parsed: ${result.successfullyParsed}`);
    console.error(`HTTP Methods: ${JSON.stringify(result.summary.byMethod, null, 2)}`);
    console.error(`With Auth: ${result.summary.withAuth}`);
    console.error(`With Rate Limiting: ${result.summary.withRateLimit}`);
    console.error(`Total Interfaces: ${result.summary.totalInterfaces}`);
  }
}

// Run main function
if (import.meta.main) {
  main().catch((error) => {
    console.error('Fatal error:', error.message);
    Deno.exit(1);
  });
}
